### ISSUE TITLE
Error getting response stream (ReadDoneAsync2): ReceiveFailure

### ISSUE THREAD

**Issue by itsBooth** [curtisshipley/CognitoForms#3]
Not sure what is causing this but every exception such as UserNotFound and NotAuthorized all trigger fine.

But when I use the correct credentials, it seems to hang for ages and then throw this exception;
Error getting response stream (ReadDoneAsync2): ReceiveFailure

Any idea what is causing it?

**Comment by curtisshipley** [comment]
I haven't seen that. I wonder if the client isn't setup correctly. If it isn't set up with SRP then it might be trying to do a redirect to a webpage.

**Comment by curtisshipley** [comment]
That's good to know.  If you give me a bit more information, I'll add it to the read me.

Also, how are things working with the code otherwise?

**Comment by kevincraig** [comment]
@itsBooth Can you provide additional information.  I'm experiencing the same error.  I didn't set up the Cognito or AWS Services so I don't know what Availability Zone was selected but can ask for it to be updated if I know what needs to be updated. Thanks.

**Comment by curtisshipley** [comment]
@kevincraig Is there an internal exception? Mind posting the stack trace?

**Comment by kevincraig** [comment]
@curtisshipley Hey just wanted to drop back by and let you know I got ours working as well. Our issue was a couple of things.  I went and looked at the pool config and we didn't have the 
`Enable sign-in API for server-based authentication (ADMIN_NO_SRP_AUTH) `
option checked as well as, we had an issue with the initial identity pool so a second identity pool was created however the roles for the pool were not recreated so I needed to create those as well.  Once I did that i was able to auth against Cognito.  I have a feeling if I blew our entire config away for Cognito and walked through your guide we'd experience no issues at this point.  Thanks for posting this and your code.  You've saved me a ton of time on this project.

**Comment by kevincraig** [comment]
@curtisshipley Just an FYI it looks as if in the latest version of Visual Studio 7.6.3 w/ Xamarin.iOS  11.14.0.14  there is a bug with AWS Cognito that is giving this error.  The one thing I did forget to mention is that I had to add my user to a manually created UserGroup that was associated with the AuthRole on the Federated UserIdentity Pool.  So we are just going to write a Lambda function for the time being on confirmation to add them to the group so they can login.   We confirmed this as one of our other devs was on an older version of Visual Studio 7.4 and he was not seeing the issue when running in the sim.  We had him upgrade to the latest VS/iOS etc and once we did that his project replicated the same behavior as above.  We're moving forward with the project but I'm trying to gather some additional info to be able to open a Bug Ticket but as of yet I'm not entirely sure who to open it with.

**Comment by curtisshipley** [comment]
@kevincraig  Thanks Kevin.  I'm surprised you had to add the `ADMIN_NO_SRP_AUTH` since some logins I've done use SRP. 

Also, thanks for the tip regarding the version issues.  Keep me apprised. 

Out of curiosity, what tool do you use to manage users on cognito? The AWS web console?

**Comment by curtisshipley** [comment]
I just tested against another cognito configuration, and the difference seems to be with setting "User Opt In" for remembering the device. If I set that to No in the cognito configuration, then I get the timeout. If I set it to "User Opt In" then I can log in -- but the refresh token code fails, probably since I'm not setting any device information in the token refresh.

**Comment by BryanGerre** [comment]
yeah I'm still having this issue even after upgrading to all the latest awssdk.  I'm glad it's not just me :) if you all have any tips I'd be happy to try.

**Comment by cjkarande** [comment]
@curtisshipley  Can you pls explain me your earlier comment on SRP?
https://github.com/curtisshipley/CognitoForms/issues/3#issuecomment-414427860

@kevincraig can you can pls elaborate on your work around using ADMIN_NO_SRP_AUTH for SignIn issue?

**Comment by BryanGerre** [comment]
@cjkarande  Let me know if you find a solution for this. I've tried everything mentioned here.

https://github.com/aws/aws-sdk-net/issues/1068

**Comment by curtisshipley** [comment]
@BryanGerre  This https://github.com/aws/aws-sdk-net/issues/1068#issuecomment-420849986 implies that if you enable the native HttpMessageHandler for each platform it works. This is an option in the build settings for Xamarin for each of Android or iOS. 

@cjkarande it may fix your issues as well.

**Comment by cjkarande** [comment]
@curtisshipley , i tried the mentioned Project options & also upgraded Core & CognitoIdentityProvider packages to latest versions as per this comment (https://github.com/aws/aws-sdk-net/issues/1068#issuecomment-434110338). However the issue is still persistent, following call in ApiCognito.cs is still getting blocked without any return value.

AuthFlowResponse context = await user.StartWithSrpAuthAsync(new InitiateSrpAuthRequest()
				{
					Password = password
				}).ConfigureAwait(false);

**Comment by curtisshipley** [comment]
@cjkarande , @BryanGerre Any luck with this?  

Reading further in that thread it looks like an HttpFactory is required, and using native HttpClients.  This is all required now since it seems that the AWS SDK requires TLS 3.0.

I've been completely swamped, and haven't had a chance to try this with the example code.

**Comment by BryanGerre** [comment]
I'll try again with the new awssdk. I've implemented the other fix mentioned in the post on forking the corrected awssdk.

**Comment by cjkarande** [comment]
Upgrading to AWSSDK.CognitoIdentityProvider 3.3.11.12 or higher & implementing HttpClientFactory in the Native Android  or iOS project fixes the issue. It worked for me for Android.

@curtisshipley, can you pls help me understand the use & flow of RefreshToken method on expiry of the Access token?

**Comment by BryanGerre** [comment]
@cjkarande can you let me know a little more about how you implemented HttpClientFactory within your native android project, and what was needed in the shared project?

**Comment by BryanGerre** [comment]
Got it figured out. I'll open a pull request for you to see the changes

**Comment by cjkarande** [comment]
@BryanGerre I have attached my changed source files for your reference.

1) I added a public static member AmazonCognitoIdentityProviderConfig to ApiCognito class
2) In MainActivity.cs I added an android implementation for HttpClientFactory
3) In MainActivity->OnCreated() I initialized the ApiCognito.AmazonCognitoIdentityProviderConfig
3) In call ApiCognito methods, i replaced 
var provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), RegionEndpoint);
with
var provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), AmazonCognitoIdentityProviderConfig);

Hope this helps you.
[ApiCognito.zip](https://github.com/curtisshipley/CognitoForms/files/2594062/ApiCognito.zip)

@curtisshipley, appreciate your reply on my earlier query about flow for RefreshToken.

**Comment by curtisshipley** [comment]
@cjkarande  The 3 AWS tokens: id token, access token and refresh token are standard oauth/JWT tokens, and follow the same conventions. This means that the access and id tokens have a specific timeout, after which then should not be used. I think the default may be an hour. 

To refresh the tokens, you have to use the Refresh Token that was given with the initial login. You make another call to InitiateAuth, but the AuthFlowType is REFRESH_TOKEN. With that call, you pass the refresh token. If successful, it will respond with a new access and id token.

Here is some more information:
https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-refresh-token

If you are using the token to access other AWS facilities, you can pass the id token and that call will be made with the appropriate privileges configured in Cognito, and Federated Identity.

You can also use the token with other servers/api calls, but you may have to validate the token yourself.  Here some information about that:

https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html
